### PR TITLE
Support fp8e5m2 dtype

### PR DESCRIPTION
**Pull request by lsy323**
Add support for `torch.fp8e5m2` dtype.

Right now `torch_xla/csrc/tensor_util.cpp` contains many duplicated codes, in the next PR I'll try to refactor it to make it cleaner.

The other fp8 variants and will be added in the following PRs

Test:
Added fp8 unit test

### CODE REVIEW

File: test/test_fp8.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import os
```
Comment (JackCaoG) [inline test/test_fp8.py:1]: should you also add this test to `test/tpu/run_test.sh`?

File: torch_xla/csrc/tensor_util.cpp [hunk 2]
```diff
@@ -62,6 +62,23 @@ struct Caster<tsl::bfloat16> {
...
+  }
+};
+
+template <>
+struct Caster<tsl::float8_e5m2> {
+  template <typename D>
+  D cast(const tsl::float8_e5m2& value) const {
+    return static_cast<D>(static_cast<float>(value));
+  }
+};
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_util.cpp:80]: what's this `Caster ` used for?

File: torch_xla/csrc/tensor_util.cpp [hunk 3]
```diff
@@ -626,6 +674,9 @@ at::Tensor MakeTensorFromXlaLiteral(const xla::Literal& literal,
     case xla::PrimitiveType::BF16:
       return XlaLiteralToTensorHelper<tsl::bfloat16>(literal,
                                                      dest_element_type);
+    case xla::PrimitiveType::F8E5M2:
```
Comment (miladm) [inline torch_xla/csrc/tensor_util.cpp:677]: qq: do you plan to bring `f8E4M3` support in a separate PR?

Comment (miladm) [reply torch_xla/csrc/tensor_util.cpp:677]: nvm - I see the PR title answers my question. thx.

**Comment by miladm** [comment]
Please include a documentation for FP8 - with example for users. Feel free to enrich the document as you add more FP8 capabilities/variants.

Side question: to my knowledge, PT AMP does not support FP8; correct?

**Comment by miladm** [comment]
@lsy323 we should include an accuracy correctness analysis in our benchmarking and documentation.

**Review by JackCaoG — APPROVED** [review]
approve to unblock, mostly lgtm we can have a follow up pr to fix review comments

File: test/test_fp8.py (hunk 1 above)
Comment (lsy323) [reply test/test_fp8.py:1]: Added

File: torch_xla/csrc/tensor_util.cpp (hunk 2 above)
Comment (lsy323) [reply torch_xla/csrc/tensor_util.cpp:80]: It's called in https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor_util.cpp#L149C6-L149C17, implementing memcpy

**Comment by lsy323** [comment]
Hi @JackCaoG, @miladm, thank you for reviewing! Sorry for the delayed respond for the comments.

> we should include an accuracy correctness analysis in our benchmarking and documentation.

In this PR the unit test is included to verify the correctness of the fp8 dtype and matmul op. Let's have documentation in a separate PR (After e4m3 variants are ready)

**Pull request merged**